clients/src/main/java/org/apache/kafka/common/record/Record.java

```diff
@@ -338,15 +338,25 @@ public int convertedSize(byte toMagic) {
      * Convert this record to another message format.
      *
      * @param toMagic The target magic version to convert to
+     * @param upconvertTimestampType The timestamp type to use if up-converting from magic 0, ignored if
+     *                               down-converting or if no conversion is needed
      * @return A new record instance with a freshly allocated ByteBuffer.
      */
-    public Record convert(byte toMagic) {
-        if (toMagic == magic())
+    public Record convert(byte toMagic, TimestampType upconvertTimestampType) {
+        byte magic = magic();
+        if (toMagic == magic)
             return this;
 
+        final TimestampType timestampType;
+        if (magic == Record.MAGIC_VALUE_V0) {
+            if (upconvertTimestampType == TimestampType.NO_TIMESTAMP_TYPE)
+                throw new IllegalArgumentException("Cannot up-convert using timestamp type " + upconvertTimestampType);
+            timestampType = upconvertTimestampType;
+        } else {
+            timestampType = TimestampType.forAttributes(attributes());
```

Comment:
Oh, I guess it's because we would never have a wrapper timestamp type if we're converting from magic 0 to magic 1, and of course it wouldn't matter when converting in the other direction. But we will need to consider it when up-converting to magic 2.